packages/components/src/box-model-overlay/index.tsx

```diff
@@ -0,0 +1,297 @@
+		// borders we added above.
+		// We want to "cancel out" those offsets by doing a `transform: translate`.
+		overlay.style.transform = `translate(calc(-50% + ${
+			( borderWidths.right - borderWidths.left ) / 2
+		}px), calc(-50% + ${
+			( borderWidths.bottom - borderWidths.top ) / 2
+		}px))`;
+
+		// Set pseudo element's position to take account for borders.
+		overlay.style.setProperty(
+			'--wp-box-model-overlay-height',
+			`${
+				domRect.height -
+				parseInt( borderTopWidth, 10 ) -
+				parseInt( borderBottomWidth, 10 )
+			}px`
+		);
+		overlay.style.setProperty(
+			'--wp-box-model-overlay-width',
+			`${
+				domRect.width -
+				parseInt( borderLeftWidth, 10 ) -
+				parseInt( borderRightWidth, 10 )
+			}px`
+		);
+		overlay.style.setProperty(
+			'--wp-box-model-overlay-top',
+			borderTopWidth
+		);
+		overlay.style.setProperty(
+			'--wp-box-model-overlay-left',
+			borderLeftWidth
+		);
+
+		// Setting padding values via CSS custom properties so that they can
+		// be applied in the pseudo elements.
+		overlay.style.setProperty(
+			'--wp-box-model-overlay-padding-top',
+			showValues.padding?.top ? paddingTop : '0'
+		);
+		overlay.style.setProperty(
+			'--wp-box-model-overlay-padding-right',
+			showValues.padding?.right ? paddingRight : '0'
+		);
+		overlay.style.setProperty(
+			'--wp-box-model-overlay-padding-bottom',
+			showValues.padding?.bottom ? paddingBottom : '0'
+		);
+		overlay.style.setProperty(
+			'--wp-box-model-overlay-padding-left',
+			showValues.padding?.left ? paddingLeft : '0'
+		);
+	}, [ targetRef, showValues.margin, showValues.padding ] );
+
+	// Make the imperative `update` method available via `ref`.
+	useImperativeHandle( ref, () => ( { update } ), [ update ] );
+
+	const getAnchorRect = useCallback(
+		() => targetRef.current.getBoundingClientRect(),
+		[ targetRef ]
+	);
+
+	// Completely skip rendering the popover if none of showValues is true.
+	const shouldShowOverlay = useMemo(
+		() =>
+			Object.values( showValues.margin ?? {} ).some(
+				( value ) => value === true
+			) ||
+			Object.values( showValues.padding ?? {} ).some(
+				( value ) => value === true
+			),
+		[ showValues.margin, showValues.padding ]
+	);
+
+	useLayoutEffect( () => {
+		const target = targetRef.current;
+
+		if ( ! shouldShowOverlay || ! target ) {
+			return;
+		}
+
+		const defaultView = target.ownerDocument.defaultView;
+
+		update();
+
+		const resizeObserver = new defaultView.ResizeObserver( update );
+		const mutationObserver = new defaultView.MutationObserver( update );
+
+		// Observing size changes.
+		resizeObserver.observe( target, { box: 'border-box' } );
+		// Observing padding and margin changes.
+		mutationObserver.observe( target, {
+			attributes: true,
+			attributeFilter: [ 'style' ],
+		} );
+
+		// Percentage paddings are based on parent element's width,
+		// so we need to also listen to the parent's size changes.
```

Comment:
This is interesting.

What about if padding and margin use VH,VW? These depend on the window sizes, are we subscribing to these changes as well.

Also for Rem unit, should we subscribe to the font size of the whole document?

I'd love if we can encapsulate that logic in a reusable hook `const { ref, padding, margin } = usePaddingMarginObserver()` it can also support a ref argument `const { padding, margin } = usePaddingMarginObserver( ref )`

Reply:
> What about if padding and margin use VH,VW? These depend on the window sizes, are we subscribing to these changes as well.

These will have to be handled by the developers manually via the `ref.current.update()` function. I don't think this can ever be done perfectly without major performance penalties (repeatedly calling `window.getComputedStyle()` in `requestAnimationFrame` for instance). I'd suggest handling it for 80% or more of the use cases by default, and also providing an escape hatch like `ref.current.update()` to handle the rest.

> I'd love if we can encapsulate that logic in a reusable hook `const { ref, padding, margin } = usePaddingMarginObserver()` it can also support a ref argument `const { padding, margin } = usePaddingMarginObserver( ref )`

I don't think returning these styles is a good idea for this hook. The returned styles will have to be passed to the overlay's inline styles. This means whenever the observer observes the changes, the styles will be updated and the overlay component will have to be re-rendered. This could be costly and unnecessary, while we can update the style attribute on the DOM directly to achieve the same effects.

In case we want to update the design of the overlay to have more complex styles, we could render a canvas and draw on it instead. We can do all this inside the hook and encapsulate them all together inside a component.

Yes, the hook might be useful somewhere else, but let's not prematurely optimize it and assume that it'd be used by some other components. We can always extract the logic into a sharable hook if we need it. Doing it prematurely means we need to take care of wider API surfaces and be careful of the backward-compatibility issues. That's why I think making it a component is the right abstraction until proven otherwise.

This is just my opinion though, as long as you think it's easier to maintain, either way works fine with me.